pkg/logql/evaluator.go

```diff
@@ -412,11 +428,15 @@ func rangeAggEvaluator(
 	if err != nil {
 		return nil, err
 	}
+	var offset int64
+	if expr.left.offset != nil {
+		offset = expr.left.offset.offset.Nanoseconds()
+	}
 	iter := newRangeVectorIterator(
 		it,
 		expr.left.interval.Nanoseconds(),
 		q.Step().Nanoseconds(),
-		q.Start().UnixNano(), q.End().UnixNano(),
+		q.Start().UnixNano(), q.End().UnixNano(), offset,
```

Comment:
I think we can simplify this as well. There's no need to make the iterator `offset` aware if we just pass `(start-offset, end-offset)` to it.

Reply:
@owen-d thanks for your comments. here we had already passed `(start-offset, end-offset)` to iterator, I am passing the offset to iterator is because I'd like to shift the returned samples timestamp from history timestamp to current timestamp, otherwise, the offset modifier might not be meaningful. Any thoughts? Please advise, thanks!

For example:
Current time is 17:00, expression - `count_over_time({job="varlogs"} [1m] offset 3h)` returns 14:00 samples, but `count_over_time({job="varlogs"} [5m])` returns 17:00 samples, then the expression - `count_over_time({job="varlogs"} [5m])/count_over_time({job="varlogs"} [5m] offset 3h) > 1.5 ` cannot work since these sub-expressions are returning inconsistent timestamp.

Reply:
I originally had the same assumption as Owen, eg. you could just pass in the modified start/end.

But now with your explanation I understand why it's like this.

I think you could only offset back L148 only and that would be fine.

https://github.com/grafana/loki/blob/05e14f995a370a3a16e7eb9d5ab37acb062c6bfc/pkg/logql/range_vector.go#L148

Reply:
thanks @cyriltovena for the correction, I noticed the same thing and made changed in https://github.com/grafana/loki/pull/3455/files#diff-f4c7f355512099579b5ea8fa4ceceb6d1eeb2b99b10233b22e0e12ce571b25d1R152 already

Reply:
I think you should add a test with an offset different than zero for the range vector iterator.

Reply:
Otherwise LGTM for me.

Reply:
make sense, let me add test, thanks @cyriltovena

Reply:
@cyriltovena added non-zero offset test - https://github.com/grafana/loki/pull/3455/commits/1b8f45407759d934622bdcedbbc795383028570a, please help review again, thanks!